metronome/checkpointing/service/test/src/io/iohk/metronome/checkpointing/service/CheckpointingServiceProps.scala

```diff
@@ -0,0 +1,407 @@
+        batch    <- genBlockChain(block, ledger, min = minChain)
+        commitQC <- genCommitQC(batch.last)
+      } yield TestFixture(block, ledger, batch, commitQC)
+    }
+
+    def genBlockChain(
+        parent: Block,
+        initialLedger: Ledger,
+        min: Int = 1,
+        max: Int = 6
+    ): Gen[List[Block]] = {
+      for {
+        n      <- Gen.choose(min, max)
+        blocks <- Gen.listOfN(n, arbitrary[Block])
+      } yield {
+        def link(
+            parent: Block,
+            prevLedger: Ledger,
+            chain: List[Block]
+        ): List[Block] = chain match {
+          case b :: bs =>
+            val nextLedger = prevLedger.update(b.body.transactions)
+            val header = b.header.copy(
+              parentHash = parent.hash,
+              height = parent.header.height + 1,
+              postStateHash = nextLedger.hash
+            )
+            val linked = Block.makeUnsafe(header, b.body)
+            linked :: link(linked, nextLedger, bs)
+          case Nil =>
+            Nil
+        }
+
+        link(parent, initialLedger, blocks)
+      }
+    }
+
+    def genCommitQC(
+        block: Block
+    ): Gen[QuorumCertificate[CheckpointingAgreement]] =
+      arbitrary[QuorumCertificate[CheckpointingAgreement]].map {
+        _.copy[CheckpointingAgreement](phase = Commit, blockHash = block.hash)
+      }
+  }
+
+  def run(fixture: TestFixture)(test: TestResources => Task[Prop]): Prop = {
+    import Scheduler.Implicits.global
+
+    fixture.resources.use(test).runSyncUnsafe(timeout = 5.seconds)
+  }
+
+  property("normal execution") = forAll(TestFixture.gen()) { fixture =>
+    run(fixture) { res =>
+      import fixture._
+      import res._
+
+      val execution = batch
+        .map(checkpointingService.executeBlock(_, commitQC, commitPath))
+        .sequence
+
+      val ledgerStorageCheck = store.runReadOnly {
+        ledgerStorage.get(finalLedger.hash)
+      }
+
+      for {
+        results         <- execution
+        persistedLedger <- ledgerStorageCheck
+        ledgerTree      <- ledgerTreeRef.get
+      } yield {
+        val ledgerTreeUpdated = ledgerTree == Map(
+          batch.last.hash -> (finalLedger, batch.last.header)
+        )
+
+        all(
+          "executed correctly" |: results.reduce(_ && _),
+          "ledger persisted" |: persistedLedger.contains(finalLedger),
+          "ledgerTree updated" |: ledgerTreeUpdated
+        )
+      }
+    }
+  }
+
+  property("failed execution - no parent") =
+    forAll(TestFixture.gen(minChain = 2)) { fixture =>
+      run(fixture) { res =>
+        import fixture._
+        import res._
+
+        // parent block or its state is not saved so this must fail
+        val execution = batch.tail
+          .map(checkpointingService.executeBlock(_, commitQC, commitPath))
+          .sequence
+
+        execution
+          .onErrorHandle(_.isInstanceOf[IllegalStateException])
+          .map(_ == true)
```

Comment:
I was starting at this for a while to understand why the last `map(_ == true)` is necessary. Maybe something like this would be clearer:
```scala
execution.attempt.map {
  case Left(ex: IllegalStateException) => ex.getMessage.contains("Could not execute block")
  case _ => false
}
```